core/src/com/google/inject/internal/FailableCache.java

```
@@ -16,65 +16,103 @@
+        errors.merge(((ErrorsException) cause).getErrors());
+      } else {
+        throw new RuntimeException(e);
+      }
+    } finally {
+      computing.remove(key);
+    }
+
+    if (errors.hasErrors()) {
       throw errors.toException();
-    } else {
-      @SuppressWarnings("unchecked") // create returned a non-error result, so this is safe
-      V result = (V) resultOrError;
-      return result;
     }
+
+    return result;
+  }
+
+  private FutureTask<V> newFutureTask(K key) {
+    return new FutureTask<>(() -> {
+      Errors errors = new Errors();
+      V result = null;
+      try {
+        result = create(key, errors);
+      } catch (ErrorsException e) {
+        errors.merge(e.getErrors());
+      }
+      if (errors.hasErrors()) throw errors.toException();
+      return result;
+    });
   }
 
   boolean remove(K key) {
-    return delegate.asMap().remove(key) != null;
+    Set<K> computing = computingThreadLocal.get();
+    if (computing.contains(key)) {
+      return false; // Don't remove a value that's still being computed.
+    }
+    synchronized (delegate) {
+      return delegate.remove(key) != null;
+    }
   }
 
   Map<K, V> asMap() {
-    return Maps.transformValues(
-        Maps.filterValues(
-            ImmutableMap.copyOf(delegate.asMap()),
-            resultOrError -> !(resultOrError instanceof Errors)),
-        resultOrError -> {
-          @SuppressWarnings("unchecked") // create returned a non-error result, so this is safe
-          V result = (V) resultOrError;
-          return result;
-        });
+    synchronized (delegate) {
+      ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+      for (Map.Entry<K, FutureTask<V>> entry : delegate.entrySet()) {
+        FutureTask<V> futureTask = entry.getValue();
+        try {
+          if (futureTask.isDone()) {
+            result.put(entry.getKey(), futureTask.get());
+          }
+        } catch (Exception ignored) {
```

"doc why its safe to ignore the exception -- namely, it was already handled above in get().  that said, it might be nice to switch this back to the old style where it returned V|Errors, which will get rid of a lot of exception handling.  it'll introduce some more casting, but IMO ditching the exception handling will end up with cleaner code even w/ the casting.  up to you."